plugins/scaffolder-node/src/actions/gitHelpers.ts

```diff
@@ -134,3 +135,71 @@ export async function commitAndPushRepo(input: {
 
   return { commitHash };
 }
+
+/**
+ * @public
+ */
+export async function commitAndPushBranch({
+  tempDir,
+  dir,
+  remoteUrl,
+  auth,
+  logger,
+  commitMessage,
+  gitAuthorInfo,
+  branch = 'master',
+  remoteRef,
+}: {
+  tempDir: string;
+  dir: string;
+  remoteUrl: string;
+  // For use cases where token has to be used with Basic Auth
+  // it has to be provided as password together with a username
+  // which may be a fixed value defined by the provider.
+  auth: { username: string; password: string } | { token: string };
+  logger: Logger;
+  commitMessage: string;
+  gitAuthorInfo?: { name?: string; email?: string };
+  branch?: string;
+  remoteRef?: string;
+}): Promise<{ commitHash: string }> {
+  const git = Git.fromAuth({
+    ...auth,
+    logger,
+  });
+
+  await git.clone({ url: remoteUrl, dir: tempDir });
```

Comment:
This doesn't really fit into the name of `commitAndPushBranch` right? 

It's doing a clone first and checking out a particular branch, I wonder if we should break this out of this function and do this seperately in the action itself instead? It might make this function a lot more re-usable.

Reply:
So, i've actually been thinking about standardizing how we interact with Git for all the upstream providers, especially when it comes to pull requests and merge requests.

Instead of favouring using the provider specific APIs to create Pull Requests and Merge Requests by providing file contents etc through the Rest API, we would basically create a branch, push the branch, and then use the provider API's to create a PR or MR from a head and base branch.

I guess this basically asks the bigger question into what is re-usable here, and what we can do in this PR to make the above suggestion a reality. I'm going to open up a followup RFC about this too, but thinking that the re-usable parts are:

- Clone a Repo
- Create a Branch
- Add an Origin (This part could be optional, thinking about permissions etc here)
- Push the Changes to the Branch

I think this is pretty much inline with the last line that you suggested, right?

Reply:
That sounds about right.  I will work on splitting that out like you mentioned and having it all tied together in the action itself.  Should have that done in a few days

Reply:
Nice! I quickly through this together too: https://github.com/backstage/backstage/issues/22244

Reply:
Changes are in.  Let me know what you think.
Also, any insight into why the E2E are failing?

Reply:
added the import and fixed your other suggestions.  I appreciate your review Fredrick